lib/datadog/di/code_tracker.rb

```diff
@@ -0,0 +1,152 @@
+        end
+
+        trace_point_lock.synchronize do
+          # Since trace point creation itself is not under a lock, see if
+          # another thread created the trace point, in which case we can
+          # disable our trace point and do nothing.
+          if @compiled_trace_point
+            # Disable the local variable, leave instance variable as it is.
+            compiled_trace_point.disable
+            return
+          end
+
+          @compiled_trace_point = compiled_trace_point
+        end
+      end
+
+      # Returns whether this code tracker has been activated and is
+      # tracking.
+      def active?
+        trace_point_lock.synchronize do
+          !!@compiled_trace_point
+        end
+      end
+
+      # Returns an array of RubVM::InstructionSequence (i.e. the compiled code)
+      # for the provided path.
+      #
+      # The argument can be a full path to a Ruby source code file or a
+      # suffix (basename + one or more directories preceding the basename).
+      # The idea with suffix matches is that file paths are likely to
+      # be different between development and production environments and
+      # the source control system uses relative paths and doesn't have
+      # absolute paths at all.
+      #
+      # Suffix matches are not guaranteed to be correct, meaning there may
+      # be multiple files with the same basename and they may all match a
+      # given suffix. In such cases, this method will return all matching
+      # paths (and all of these paths will be attempted to be instrumented
+      # by upstream code).
+      #
+      # If the suffix matches one of the paths completely (which requires it
+      # to be an absolute path), only the exactly matching path is returned.
+      # Otherwise all known paths that end in the suffix are returned.
+      # If no paths match, an empty array is returned.
+      def iseqs_for_path(suffix)
+        registry_lock.synchronize do
+          exact = registry[suffix]
+          if exact
+            return [exact]
+          end
+          inexact = []
+          registry.each do |path, iseq|
+            # Exact match is not possible here, meaning any matching path
+            # has to be longer than the suffix. Require full component matches,
+            # meaning either the first character of the suffix is a slash
+            # or the previous character in the path is a slash.
+            # For now only check for forward slashes for Unix-like OSes;
+            # backslash is a legitimate character of a file name in Unix
+            # therefore simply permitting forward or back slash is not
+            # sufficient, we need to perform an OS check to know which
+            # path separator to use.
+            if path.length > suffix.length && (
+              path[path.length - suffix.length - 1] == "/" ||
+              suffix[0] == "/"
+            ) && path.end_with?(suffix)
+              inexact << iseq
+            end
```

Comment:
I think something like this. Not sure where do we stand in "comments vs code", but maybe this may be it?

```suggestion
            if path.length > suffix.length
              previous_char = path[path.length - suffix.length - 1]
              inexact << iseq if previous_char == "/" && path.end_with?(suffix)
            end
```

Reply:
It looks to me like you lost the `suffix[0] == "/"` condition along the way and I personally find your proposed version more difficult to follow because it lacks symmetry where there is symmetry in the logic and inverts cause and effect with the postfix if, thus also putting the action in the middle of the conditions, but I'm happy to commit a rendering you are happy with as long as it is equivalent functionally to what is currently in the diff.

Reply:
Sorry, diff was written under high distraction, you're right, added `suffix` check.

```suggestion
            if path.length > suffix.length && path.end_with?(suffix)
              previous_char = path[path.length - suffix.length - 1]

              inexact << iseq if (previous_char == "/" || suffix[0] == "/")
            end
```

How about this variant. I've grouped the path conditions which were guarding the whole `if` statement anyway. I understand that it might exit later than your statement, but now I think it's more readable (I know that's a matter of taste and perception sometimes, so no hard feelings)